pandaserver/dataservice/setupper.py

```diff
@@ -0,0 +1,207 @@
+                    # use ATLAS plug-in by default
+                    from pandaserver.dataservice.setupper_atlas_plugin import (
+                        SetupperAtlasPlugin,
+                    )
+
+                    setupper_plugin_class = SetupperAtlasPlugin
+                tmp_log.debug(f"plugin name -> {setupper_plugin_class.__name__}")
+                try:
+                    # make plugin
+                    setupper_plugin = setupper_plugin_class(
+                        self.task_buffer,
+                        self.jobs,
+                        tmp_log,
+                        resubmit=self.resubmit,
+                        panda_ddm=self.panda_ddm,
+                        ddm_attempt=self.ddm_attempt,
+                        only_ta=self.only_ta,
+                        first_submission=self.first_submission,
+                    )
+                    # run plugin
+                    tmp_log.debug("run plugin")
+                    setupper_plugin.run()
+                    # go forward if not TA
+                    if not self.only_ta:
+                        # update jobs
+                        tmp_log.debug("update jobs")
+                        self.update_jobs(setupper_plugin.jobs + setupper_plugin.jumbo_jobs, tmp_log)
+                        # execute post process
+                        tmp_log.debug("post execute plugin")
+                        setupper_plugin.post_run()
+                    tmp_log.debug("done plugin")
+                except Exception:
+                    error_type, error_value = sys.exc_info()[:2]
+                    tmp_log.error(f"plugin failed with {error_type}:{error_value}")
+            tmp_log.debug("main end")
+        except Exception as error:
+            tmp_log.error(f"master failed with {str(error)} {traceback.format_exc()}")
+
+    #  update jobs
+    def update_jobs(self, job_list: List[object], tmp_log: LogWrapper) -> None:
+        """
+        Updates the status of jobs.
+
+        :param job_list: The list of jobs to be updated.
+        :param log: The logger to be used for logging.
+        """
+        update_jobs = []
+        failed_jobs = []
+        activate_jobs = []
+        waiting_jobs = []
+        # sort out jobs
+        for job in job_list:
+            # failed jobs
+            if job.jobStatus in ["failed", "cancelled"]:
+                failed_jobs.append(job)
+            # waiting
+            elif job.jobStatus == "waiting":
+                waiting_jobs.append(job)
+            # no input jobs
+            elif job.dispatchDBlock == "NULL":
+                activate_jobs.append(job)
+            # normal jobs
+            else:
+                # change status
+                job.jobStatus = "assigned"
+                update_jobs.append(job)
+        # trigger merge generation if all events are done
+        new_activate_jobs = []
+        n_finished = 0
+        for job in activate_jobs:
```

Comment:
I don't understand the logic why an activate (=activated?) job suddenly is updated as finished, what is the notDiscardEvents() or allOkEvents()?
What is the False in updateJobs([job], False)?

Reply:
same for the next block. This logic needs more comments to indicate how this is working